### PR TITLE
Refactor for scraped

### PR DESCRIPTION
**Pull request by ondenman**
Part of: https://github.com/everypolitician/everypolitician-data/issues/25297

### CODE REVIEW

File: lib/decorators/party_name_attributes.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    'independiente.png'             => ['Sin Partido', '_IND'],
+  }.freeze
+
+  def body
+    Nokogiri::HTML(super).tap do |doc|
+      doc.css('img').each do |img|
+        key = img.attr('src').split('/').last
+        next unless PARTY_IMAGES_MAPPED.key?(key)
+        img[:party_id] = PARTY_IMAGES_MAPPED[key].last
+        img[:party_name] = PARTY_IMAGES_MAPPED[key].first
```
Comment (tmtmtmtm) [inline lib/decorators/party_name_attributes.rb]: Checking if there _is_ a key and then looking up by it is a little bit repetitive. I'd suggest something like this would be a little bit clearer and simpler:

```ruby
        party_data = PARTY_IMAGES_MAPPED[key] or next
        img[:party_name] = party_data.first
        img[:party_id] = party_data.last
```

(though `PARTY_IMAGES_MAPPED` is a slightly odd name. `PARTY_IMAGE_MAP` would be clearer too IMO)

File: lib/decorators/party_name_attributes.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    'LogoMorena.jpg'                => ['Movimiento Regeneración Nacional', 'MORENA'],
+    'encuentro.png'                 => ['Encuentro Social', 'PES'],
+    'logo_SP.jpg'                   => ['Sin Partido', '_IND'],
+    'independiente.png'             => ['Sin Partido', '_IND'],
+  }.freeze
+
+  def body
+    Nokogiri::HTML(super).tap do |doc|
+      doc.css('img').each do |img|
+        key = img.attr('src').split('/').last
```
Comment (tmtmtmtm) [inline lib/decorators/party_name_attributes.rb]: `File.basename(img[:src])` would probably be clearer and simpler here.

File: lib/members_page.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      noko.xpath('td/a/@href').text
+    end
+
+    field :area do
+      "#{noko.xpath('td')[1].text} #{district.split(' ').last}"
+    end
+
+    field :area_id do
+      area_part = area.split(' ').reject { |x| x =~ /\d/ }.join('-').downcase
+      "ocd-division/country:mx/entidad:#{area_part}/#{d_c_no}"
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: There are two things happening here of different types, and it seems odd to have `d_c_no` as (which could really do with a better name) as its own method, and not `area_part` (likewise)

File: lib/members_page.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    field :sort_name do
+      noko.xpath('td').first.text.tidy.sub(/^\d+\s*/, '').sub(' (LICENCIA)', '')
+    end
+
+    field :source do
+      noko.xpath('td/a/@href').text
+    end
+
+    field :area do
+      "#{noko.xpath('td')[1].text} #{district.split(' ').last}"
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: It would improve readability a lot, I suspect, if these two things were split out into named methods.

File: test/members_page_test.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      members_page.party_groupings.first.members.first
+    end
+
+    it 'should have the expected data' do
+      subject.to_h.must_equal(
+        id:        '260',
+        sort_name: 'Abdala Carmona Yahleel',
+        source:    'http://sitl.diputados.gob.mx/LXIII_leg/curricula.php?dipt=260',
+        area:      'Tamaulipas 1',
+        area_id:   'ocd-division/country:mx/entidad:tamaulipas/distrito:1'
```
Comment (tmtmtmtm) [inline test/members_page_test.rb]: As most of the complex logic is about the two different types of areas, I think we really need a test for the other as well.

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]

File: scraper.rb [hunk 6]
```diff
@@ -1,95 +1,37 @@
...
 require 'scraped'
 require 'scraperwiki'
 
 # require 'open-uri/cached'
 # OpenURI::Cache.cache_path = '.cache'
 require 'scraped_page_archive/open-uri'
 
+require_rel 'lib'
+
 def noko_for(url)
```
Comment (tmtmtmtm) [inline scraper.rb]: refactoring to use `Scraped` usually means getting rid of helpers like this…

File: test/members_page_test.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  let :members_page do
+    url = 'http://sitl.diputados.gob.mx/LXIII_leg/listado_diputados_gpnp.php?tipot=TOTAL'
+    MembersPage.new(response: Scraped::Request.new(url: url).response)
+  end
+
+  subject do
+    party = members_page.party_groupings.find do |party_group|
+      party_group.party_id == party_id
+    end
+    party.members.find { |mem| mem.id == id }
```
Comment (tmtmtmtm) [inline test/members_page_test.rb]: This is a pretty complex set-up. If member IDs are unique, then this could all be simplified to 

```ruby
  subject do
    members_page.party_groupings.flat_map(&:members).find { |mem| mem.id == id }
  end 
```
Then you could drop the necessity for each test to set up *two* identifiers — the :id would be sufficient.

File: test/members_page_test.rb [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  end
+
+  describe 'party groupings' do
+    subject do
```
Comment (tmtmtmtm) [inline test/members_page_test.rb]: This is clobbering an outer `subject`. If these are two different types of tests, they should be separately nested.

File: lib/members_page.rb [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+    def area_part
+      area.split(' ').reject { |x| x =~ /\d/ }.join('-').downcase
+    end
+
+    def district
+      noko.xpath('td')[2].text
+    end
+
+    def district_number
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: This is quite a misleading name, as it doesn't only apply to districts.

File: lib/members_page.rb [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+      "ocd-division/country:mx/entidad:#{area_part}/#{formatted_district_number}"
+    end
+
+    private
+
+    def area_name
+      noko.xpath('td')[1].text
+    end
+
+    def area_part
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: This is a confusing/misleading name. What does "part" mean here?

File: lib/members_page.rb [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+    def district
+      noko.xpath('td')[2].text
+    end
+
+    def district_number
+      district.split(' ').last
+    end
+
+    def formatted_district_number
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: as well as being misleading by implying it's only about districts, saying "formatted" doesn't really convey much here. Anyone reading this is instantly going to say "Formatted *how*?"

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]
@ondenman I'm getting the impression that you don't actually follow what's going on with generating the area data here. I'd suggest taking the time to understand what we're doing, and then picking good names that will help others understand it too, rather than obfuscating matters more.

(Tip: your helper methods should clearly distinguish between source data, and us formatting that in specific ways)

File: lib/decorators/party_name_attributes.rb (hunk 2 above)
Comment (tmtmtmtm) [reply lib/decorators/party_name_attributes.rb]: @ondenman you haven't addressed this comment, either by changing the code, or leaving a message to explain why…

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]
@ondenman when marking things for re-review, please ensure that all issues previously raised are addressed or discussed. (I haven't yet reviewed the changes, due to this)

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]
I'm afraid these new names are still quite misleading as before. Can you trying explaining in English what is happening here, before trying to implement it in code?

File: .rubocop-https---raw-githubusercontent-com-everypolitician-everypolitician-data-master--rubocop-base-yml [hunk 12]
```diff
@@ -0,0 +1,59 @@
+# A base Rubocop configuration that other repos can inherit from
```
Comment (tmtmtmtm) [inline .rubocop-https---raw-githubusercontent-com-everypolitician-everypolitician-data-master--rubocop-base-yml]: Why is this file here? This seems like you've just done a bulk add and commit without checking what's going in…

File: scraper.rb [hunk 13]
```diff
@@ -1,95 +1,31 @@
...
-    if tds.size == 1
-      if img = tr.css('img/@src').text
-        next if img.to_s.empty?
-        next if img == 'images/h_line.gif'
-        next if img == 'images/lin_por_gp.jpg'
-        party, party_id = party_info_for(img)
-        warn party.to_s
-      end
-      next
+def scrape_list(term, url)
```
Comment (tmtmtmtm) [inline scraper.rb]: Now that you've changed this from a command method to a query method, you should probably change the name to be less command-y. (Yeah, `scrape` suffers from the same problem, but we plan to get rid of that with the `Scraper` class)

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]
Naming is much better now!

File: .rubocop-https---raw-githubusercontent-com-everypolitician-everypolitician-data-master--rubocop-base-yml (hunk 12 above)
Comment (ondenman) [reply .rubocop-https---raw-githubusercontent-com-everypolitician-everypolitician-data-master--rubocop-base-yml]: I've removed this in a fix up: 1ce5f81098b5b55a9984fca3bc44393b1432eb36

File: scraper.rb (hunk 13 above)
Comment (ondenman) [reply scraper.rb]: I've renamed this in: ce2fc501dfe0a9e6dd2d18dd3d778e85b61646d4

**Comment by tmtmtmtm** [comment]
@ondenman autosquash doesn't apply cleanly here. Can you rebase and force-push?

**Review by tmtmtmtm — COMMENTED** [review]
Woohoo. We'll want to evolve this further later, particularly wrt to the tests, but 👍

**Review by tmtmtmtm — APPROVED** [review]
